dns/resolver.go

```diff
@@ -0,0 +1,30 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dns
+
+/*
+	A resolver is a function that resolves a hostname to a list of IP addresses.
+
+	TODO: provide factory methods to create a Resolver. For example:
+
+	```go
+		manualDns := makeManualResolver(
+			map[string][]string{ "example.com": []string{"255.255.255.255:80" }
+		})	
+	
+		dialer := &shadowsocks.NewStreamDialer(endpoint, key, manualDns)
+	````
+*/
+type Resolver func(host string) ([]string, error)
```

Comment:
A function is unmodifiable. Think about the following scenario:

```go
dohResolver := NewDoHResolver(...)
dialer := NewDialerWithResolver(dohResolver)
// ... initially
dohResolver.SetTimeout(xxx)
// ... after some time
dohResolver.SetTimeout(yyy)
```

Unless we decide that all our Resolver implementations are immutable, I would prefer to define it as an interface: 

```
type Resolver interface {
  LookupIP func(string)([]string, error)
}
```

Reply:
For example Intra allows to you change the DoH server on the fly (without disconnecting and connecting again).

Reply:
Sure it does. But it means we will create a new `interface` in the app that wraps around this `func`.

Reply:
Well yes, define a `struct` that wraps around the function, which is OK for a single app.

My concern is whether a "stateful" `Resolver` is common across all apps, and how it interacts with `StreamDialer` and `PacketProxy`.

For example, some implementation might maintain a connection to a remote server, then we need to have `Close()` function in the `Resolver` to prevent resource leak. 

We can discuss them in the brainstorm meeting, it all depends on the use cases we would like to cover. We might need to consider more stuffs than the `LocalResolver` or `shadowsocks.StreamDialer`.

Another thing I can think of is the two `Resolve` methods, one with `context.Context` and one without (similar to Go [`Dialer`'s `Dial` and `DialContext`](https://pkg.go.dev/net#Dialer.Dial))? We can discuss whether we should support it too.